src/compiler/checker.ts

```diff
@@ -6807,7 +6807,7 @@ namespace ts {
 
                     // Record a new minimum argument count if this is not an optional parameter
                     const isOptionalParameter = param.initializer || param.questionToken || param.dotDotDotToken ||
-                        iife && parameters.length > iife.arguments.length && !param.type ||
+                        iife && parameters.length > iife.arguments.length && !param.type || param.type && param.type.kind === SyntaxKind.VoidKeyword ||

```

Comment:
A real fix here is quite a bit more involved, unfortunately.

This change doesn't fix the linked issue because it's only looking for the exact *syntactic* form of `function f(x: void)` (which no one would write) instead of actually detecting the type of the parameter during overload resolution and treating it as optional.